gtest/src/manager.rs

```diff
@@ -160,25 +169,644 @@ impl ExtManager {
+
+        let _ = self.task_pool.delete(
+            slot.finish,
+            ScheduledTask::RemoveGasReservation(program_id, reservation),
+        );
+    }
+
+    pub(crate) fn remove_gas_reservation_slot(
+        &mut self,
+        reservation: ReservationId,
+        slot: GasReservationSlot,
+    ) -> GasReservationSlot {
+        let interval = Interval {
+            start: slot.start,
+            finish: slot.finish,
+        };
+
+        self.charge_for_hold(reservation, interval, StorageType::Reservation);
+        self.consume_and_retrieve(reservation);
+
+        slot
+    }
+
+    pub(crate) fn consume_and_retrieve(&mut self, id: impl Origin) {
+        let outcome = self.gas_tree.consume(id).unwrap_or_else(|e| {
+            let err_msg = format!(
+                "consume_and_retrieve: failed consuming the rest of gas. Got error - {e:?}"
+            );
+
+            log::error!("{err_msg}");
+            unreachable!("{err_msg}")
+        });
+
+        if let Some((imbalance, multiplier, external)) = outcome {
+            let gas_left = imbalance.peek();
+
+            if !gas_left.is_zero() {
+                self.bank
+                    .withdraw_gas(external.cast(), gas_left, multiplier);
+            }
+        }
+    }
+
     /// Insert message into the delayed queue.
-    fn send_delayed_dispatch(&mut self, dispatch: Dispatch, delay: u32) {
+    pub(crate) fn send_delayed_dispatch(
+        &mut self,
+        origin_msg: MessageId,
+        dispatch: Dispatch,
+        delay: u32,
+        to_user: bool,
+        reservation: Option<ReservationId>,
+    ) {
+        if delay.is_zero() {
+            let err_msg = "send_delayed_dispatch: delayed sending with zero delay appeared";
+
+            log::error!("{err_msg}");
+            unreachable!("{err_msg}");
+        }
+
         let message_id = dispatch.id();
-        let task = if Actors::is_program(dispatch.destination()) {
-            ScheduledTask::SendDispatch(message_id)
+
+        if self.dispatches_stash.contains_key(&message_id) {
+            let err_msg = format!(
+                "send_delayed_dispatch: stash already has the message id - {id}",
+                id = dispatch.id()
+            );
+
+            log::error!("{err_msg}");
+            unreachable!("{err_msg}");
+        }
+
+        // Validating dispatch wasn't sent from system with delay.
+        if dispatch.is_error_reply() || matches!(dispatch.kind(), DispatchKind::Signal) {
+            let err_msg = format!(
+                "send_delayed_dispatch: message of an invalid kind is sent: {kind:?}",
+                kind = dispatch.kind()
+            );
+
+            log::error!("{err_msg}");
+            unreachable!("{err_msg}");
+        }
+
+        let mut to_mailbox = false;
+
+        let sender_node = reservation
+            .map(|id| id.cast::<PlainNodeId>())
+            .unwrap_or_else(|| origin_msg.cast::<PlainNodeId>())
+            .cast::<MessageId>();
```

Comment:
you can convert it into origin using `into_origin`, instead of casting to `MessageId`, as it looks strange to cast reservation to message type.